include/veriblock/blockchain/block_index.hpp

```diff
@@ -43,21 +43,42 @@ struct BlockIndex : public Block::addon_t {
    * ancestors are connected.
    * @return true if block is connected, false otherwise.
    */
-  bool isConnected() const noexcept { return this->hasFlags(BLOCK_CONNECTED); }
+  bool isConnected() const noexcept {
+    return this->isValidUpTo(BLOCK_CONNECTED);
+  }
 
   uint32_t getStatus() const { return status; }
   void setStatus(uint32_t _status) {
     this->status = _status;
     setDirty();
   }
 
+  uint32_t getValidityLevel() const {
+    auto level = status & BLOCK_VALID_MASK;
+    VBK_ASSERT_MSG(level <= BLOCK_CAN_BE_APPLIED, "unknown validity level");
+    return level;
+  }
+
   bool isValid(enum BlockStatus upTo = BLOCK_VALID_TREE) const {
-    VBK_ASSERT(!(upTo & ~BLOCK_VALID_MASK));  // Only validity flags allowed.
     if ((status & BLOCK_FAILED_MASK) != 0u) {
       // block failed
       return false;
     }
-    return ((status & BLOCK_VALID_MASK) >= upTo);
+    return isValidUpTo(upTo);
+  }
+
+  bool isValidUpTo(enum BlockStatus upTo) const {
+    VBK_ASSERT_MSG(!(upTo & ~BLOCK_VALID_MASK),
+                   "Only validity flags are allowed");
+
+    auto validityLevel = getValidityLevel();
+    VBK_ASSERT_MSG(
+        validityLevel != BLOCK_CAN_BE_APPLIED || !hasFlags(BLOCK_FAILED_POP),
```

Comment:
Here also can simplify with:
!(validityLevel == BLOCK_CAN_BE_APPLIED && hasFlags(BLOCK_FAILED_POP))
that is also consistent with the comment below